core/src/main/scala/kafka/server/FetchSession.scala

```diff
@@ -0,0 +1,690 @@
+                      var highWatermark: Long,
+                      var fetcherLogStartOffset: Long,
+                      var localLogStartOffset: Long)
+    extends ImplicitLinkedHashSet.Element {
+
+  var cachedNext: Int = ImplicitLinkedHashSet.INVALID_INDEX
+  var cachedPrev: Int = ImplicitLinkedHashSet.INVALID_INDEX
+
+  override def next = cachedNext
+  override def setNext(next: Int) = this.cachedNext = next
+  override def prev = cachedPrev
+  override def setPrev(prev: Int) = this.cachedPrev = prev
+
+  def this(topic: String, partition: Int) =
+    this(topic, partition, -1, -1, -1, -1, -1)
+
+  def this(part: TopicPartition) =
+    this(part.topic(), part.partition())
+
+  def this(part: TopicPartition, reqData: FetchRequest.PartitionData) =
+    this(part.topic(), part.partition(),
+      reqData.maxBytes, reqData.fetchOffset, -1,
+      reqData.logStartOffset, -1)
+
+  def this(part: TopicPartition, reqData: FetchRequest.PartitionData,
+           respData: FetchResponse.PartitionData) =
+    this(part.topic(), part.partition(),
+      reqData.maxBytes, reqData.fetchOffset, respData.highWatermark,
+      reqData.logStartOffset, respData.logStartOffset)
+
+  def topicPartition() = new TopicPartition(topic, partition)
+
+  def reqData() = new FetchRequest.PartitionData(fetchOffset, fetcherLogStartOffset, maxBytes)
+
+  def updateRequestParams(reqData: FetchRequest.PartitionData): Unit = {
+    // Update our cached request parameters.
+    maxBytes = reqData.maxBytes
+    fetchOffset = reqData.fetchOffset
+    fetcherLogStartOffset = reqData.logStartOffset
+  }
+
+  /**
+    * Update this CachedPartition with new request and response data.
+    *
+    * This function should be called while holding the appropriate CachedPartitionMap
+    * lock.
+    *
+    * @return True if this partition should be included in the FetchResponse
+    *         we send back to the fetcher; false if it can be omitted.
+    */
+  def updateResponseData(respData: FetchResponse.PartitionData): Boolean = {
+    // Check the response data.
+    var mustRespond = false
+    if ((respData.records != null) && (respData.records.sizeInBytes() > 0)) {
+      // Partitions with new data are always included in the response.
+      mustRespond = true
+    }
+    if (highWatermark != respData.highWatermark) {
+      mustRespond = true
+      highWatermark = respData.lastStableOffset
+    }
+    if (localLogStartOffset != respData.logStartOffset) {
+      mustRespond = true
+      localLogStartOffset = respData.logStartOffset
+    }
+    if (respData.error.code() != 0) {
+      // Partitions with errors are always included in the response.
+      // We also set the cached lastStableOffset to an invalid offset, -1.
+      // This ensures that when the error goes away, we re-send the partition.
+      highWatermark = -1
+      mustRespond = true
+    }
+    mustRespond
+  }
+
+  override def hashCode() = Objects.hash(topic, new Integer(partition))
```

Comment:
The fact that we are allocating an array and potentially an `Integer` to compute the hashCode is suboptimal given that these elements are meant to be added to the `ImplicitLinkedHashSet`, which doesn't seem to cache hash codes.